### ISSUE TITLE
Any plans for a QUE instrumentation?

### ISSUE THREAD

**Issue by hs-bguven** [DataDog/dd-trace-rb#1141]
Any plans on introducing instrumentation for [QUE](https://github.com/que-rb/que)? 

It's one of the last critical pieces remaining to get distributed tracing in our system. 

I glanced at other instrumentations like sidekiq (which I assumed would be the most similar to `que`) to see if I could create a PR but it wasn't the most obvious to me out of the gate on how to do it.

I would be more than happy to give it a shot with some help/direction.

**Comment by marcotc** [comment]
Hey @hs-bguven, thank you for reaching out!

We haven't received other request for `que` support, but we can still help unblock you on anything you need if you want to create a pull request with that feature.

First thing we'd need to know: what version of `que` are you interested in? I noticed `0.14` hasn't received any updates in the last few years, while `1.0` has, albeit in beta state.

The main reason I ask is because version `1.0` has [middleware support](https://github.com/que-rb/que/blob/ad5f02cc1ae68c00699d04da8c9c8e5b4262c5a2/lib/que/utils/middleware.spec.rb#L15-L34), which makes creating a `ddtrace` integration super easy.

If you are interested in hooking into `< 1.0` is still possible, but will require monkey-patching the correct methods for specific versions of the gem.

If we are going with the middleware implementation, the main instrumentation code would look something like this:
```ruby
Que.job_middleware.push(-> (job, &block) do
  @tracer.trace(Ext::SPAN_JOB, service: configuration[:service_name], span_type: Datadog::Ext::AppTypes::WORKER) do |span|
    span.resource = extract_resource(job)

    # Any relevant fields we can find in the `job` object.
    # These fields should:
    # * Provide reasonable value to the developer
    # * Not contain personally identifiable information (on a reasonable deployment)
    span.set_tag(Ext::TAG_JOB_QUEUE, job.queue)
    span.set_tag(Ext::TAG_JOB_PRIORITY, job.priority)

    # ddtrace internal: Set analytics sample rate
    if Contrib::Analytics.enabled?(configuration[:analytics_enabled])
      Contrib::Analytics.set_sample_rate(span, configuration[:analytics_sample_rate])
    end

    # ddtrace internal: Measure service stats
    Contrib::Analytics.set_measured(span)

    block.call # Continue Que processing pipeline
  end
end
)
```

The **best** reference I can find for a community-driven gem instrumentation is the recent work on the `sneakers` gem, merged just a few weeks ago: https://github.com/DataDog/dd-trace-rb/pull/1121
It is also a background worker gem, so you'll see a lot of similarities. That PR has a minimal and complete example of a new gem integration.

Let us know if this information is clear so far, and what we can do to support you with this contribution!

**Comment by hs-bguven** [comment]
Thank you for your quick reply @marcotc.

We are currently using que@1.0. Having seen the middleware ability in their readme gave me the confidence in the first place : )

I have taken a look at the sneaker implementation and with the example you have provided above, I think I can give it a try.

I am going to try and test it out over the next few days as I can carve out time. Fingers crossed : )

Thanks again for the example and the reference links.

**Comment by marcotc** [comment]
👋 @hs-bguven, your PR is has now been released, as part of [0.40.0](https://github.com/DataDog/dd-trace-rb/releases/tag/v0.40.0)!

Let us know if there's any further feedback on it. I'm marking this issue as resolved now.

**Comment by hs-bguven** [comment]
@marcotc Thanks a lot. 

We have been using it in prod without any issues. I have a question or two about improvements but I will create a separate issue ticket to keep track of it.
